File: sdks/go/wdk/multisig/signature.go
package multisig

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"sort"
	"sync"
	"time"

	"github.com/ethereum/go-ethereum/common"
)

// SignatureCollector manages pending multi-sig transactions and signature collection.
type SignatureCollector struct {
	mu              sync.RWMutex
	pendingRequests map[string]*TransactionRequest
	expirationSecs  int64
}

// SignatureCollectorConfig holds configuration for the collector.
type SignatureCollectorConfig struct {
	// ExpirationSeconds is the request expiration time (default: 3600)
	ExpirationSeconds int64
}

// NewSignatureCollector creates a new signature collector.
func NewSignatureCollector(config *SignatureCollectorConfig) *SignatureCollector {
	expSecs := int64(DefaultRequestExpirationSeconds)
	if config != nil && config.ExpirationSeconds > 0 {
		expSecs = config.ExpirationSeconds
	}

	return &SignatureCollector{
		pendingRequests: make(map[string]*TransactionRequest),
		expirationSecs:  expSecs,
	}
}

// CreateRequest creates a new signature collection request.
func (c *SignatureCollector) CreateRequest(
	safeAddress common.Address,
	tx *SafeTransaction,
	txHash common.Hash,
	owners []common.Address,
	threshold int,
) *TransactionRequest {
	c.mu.Lock()
	defer c.mu.Unlock()

	now := currentTimestamp()
	request := &TransactionRequest{
		ID:              generateRequestID(),
		SafeAddress:     safeAddress,
		Transaction:     tx,
		TransactionHash: txHash,
		Signatures:      make(map[common.Address]*SafeSignature),
		Threshold:       threshold,
		CreatedAt:       now,
		ExpiresAt:       now + c.expirationSecs,
	}

	c.pendingRequests[request.ID] = request
	return request
}

// AddSignature adds a signature to a request.
func (c *SignatureCollector) AddSignature(requestID string, sig *SafeSignature) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	request, exists := c.pendingRequests[requestID]
	if !exists {
		return errors.New("request not found")
	}

	// Check expiration
	if currentTimestamp() > request.ExpiresAt {
		delete(c.pendingRequests, requestID)
		return errors.New("request expired")
	}

	// Check if already signed by this signer
	if _, exists := request.Signatures[sig.Signer]; exists {
		return errors.New("already signed by this signer")
	}

	request.Signatures[sig.Signer] = sig
	return nil
}

// GetRequest retrieves a pending request.
func (c *SignatureCollector) GetRequest(requestID string) (*TransactionRequest, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	request, exists := c.pendingRequests[requestID]
	if !exists {
		return nil, errors.New("request not found")
	}

	// Check expiration
	if currentTimestamp() > request.ExpiresAt {
		return nil, errors.New("request expired")
	}

	return request, nil
}

// RemoveRequest removes a request.
func (c *SignatureCollector) RemoveRequest(requestID string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.pendingRequests, requestID)
}

// GetPendingRequests returns all non-expired pending requests.
func (c *SignatureCollector) GetPendingRequests() []*TransactionRequest {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.cleanupExpiredLocked()

	var requests []*TransactionRequest
	for _, request := range c.pendingRequests {
		requests = append(requests, request)
	}
	return requests
}

// GetPendingOwners returns owners who haven't signed a request yet.
func (c *SignatureCollector) GetPendingOwners(requestID string, owners []common.Address) ([]common.Address, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	request, exists := c.pendingRequests[requestID]
	if !exists {
		return nil, errors.New("request not found")
	}

	var pending []common.Address
	for _, owner := range owners {
		if _, signed := request.Signatures[owner]; !signed {
			pending = append(pending, owner)
		}
	}
	return pending, nil
}

// GetSignedOwners returns owners who have signed a request.
func (c *SignatureCollector) GetSignedOwners(requestID string) ([]common.Address, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	request, exists := c.pendingRequests[requestID]
	if !exists {
		return nil, errors.New("request not found")
	}

	var signed []common.Address
	for signer := range request.Signatures {
		signed = append(signed, signer)
	}
	return signed, nil
}

// GetCombinedSignature returns the packed signatures for execution.
func (c *SignatureCollector) GetCombinedSignature(requestID string) ([]byte, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	request, exists := c.pendingRequests[requestID]
	if !exists {
		return nil, errors.New("request not found")
	}

	if !request.IsReady() {
		return nil, errors.New("not enough signatures")
	}

	// Sort signers by address
	var signers []common.Address
	for signer := range request.Signatures {
		signers = append(signers, signer)
	}
	sortAddresses(signers)

	// Pack signatures
	var packed []byte
	for _, signer := range signers {
		sig := request.Signatures[signer]
		packed = append(packed, sig.Signature...)
	}

	return packed, nil
}

// Cleanup removes expired requests.
func (c *SignatureCollector) Cleanup() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cleanupExpiredLocked()
}

// cleanupExpiredLocked removes expired requests (must hold lock).
func (c *SignatureCollector) cleanupExpiredLocked() {
	now := currentTimestamp()
	for id, request := range c.pendingRequests {
		if now > request.ExpiresAt {
			delete(c.pendingRequests, id)
		}
	}
}

// Clear removes all pending requests.
func (c *SignatureCollector) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.pendingRequests = make(map[string]*TransactionRequest)
}

// Helper functions

// generateRequestID creates a unique request identifier.
func generateRequestID() string {
	timestamp := time.Now().UnixMilli()
	randomBytes := make([]byte, 4)
	rand.Read(randomBytes)
	return "msig_" + hex.EncodeToString([]byte{
		byte(timestamp >> 24), byte(timestamp >> 16),
		byte(timestamp >> 8), byte(timestamp),
	}) + "_" + hex.EncodeToString(randomBytes)
}

// currentTimestamp returns the current Unix timestamp in seconds.
func currentTimestamp() int64 {
	return time.Now().Unix()
}

// sortAddresses sorts addresses in ascending order.
func sortAddresses(addrs []common.Address) {
	sort.Slice(addrs, func(i, j int) bool {
		return addrs[i].Hex() < addrs[j].Hex()
	})
}

// CombineSignatures combines multiple signatures sorted by signer address.
func CombineSignatures(sigs map[common.Address]*SafeSignature) []byte {
	var signers []common.Address
	for signer := range sigs {
		signers = append(signers, signer)
	}
	sortAddresses(signers)

	var packed []byte
	for _, signer := range signers {
		sig := sigs[signer]
		packed = append(packed, sig.Signature...)
	}

	return packed
}

// IsValidThreshold checks if a threshold is valid for the given owner count.
func IsValidThreshold(threshold, ownerCount int) bool {
	return threshold >= MinThreshold && threshold <= ownerCount
}

// AreAddressesUnique checks if all addresses are unique.
func AreAddressesUnique(addrs []common.Address) bool {
	seen := make(map[common.Address]struct{})
	for _, addr := range addrs {
		if _, exists := seen[addr]; exists {
			return false
		}
		seen[addr] = struct{}{}
	}
	return true
}

// GetOwnerIndex returns the index of an owner in the list, or -1 if not found.
func GetOwnerIndex(owner common.Address, owners []common.Address) int {
	for i, o := range owners {
		if o == owner {
			return i
		}
	}
	return -1
}

File: sdks/python/t402/src/t402/multisig/signature.py
"""
Signature collector for T402 Multi-sig support.
"""

import threading
from typing import Dict, List, Optional

from .constants import DEFAULT_REQUEST_EXPIRATION_SECONDS
from .types import (
    SafeSignature,
    SafeTransaction,
    TransactionRequest,
)
from .utils import generate_request_id, current_timestamp, sort_addresses, combine_signatures


class SignatureCollector:
    """Manages pending multi-sig transactions and signature collection."""

    def __init__(self, expiration_seconds: Optional[int] = None):
        """
        Initialize SignatureCollector.

        Args:
            expiration_seconds: Request expiration time (default: 1 hour).
        """
        self._pending_requests: Dict[str, TransactionRequest] = {}
        self._expiration_seconds = (
            expiration_seconds or DEFAULT_REQUEST_EXPIRATION_SECONDS
        )
        self._lock = threading.RLock()

    def create_request(
        self,
        safe_address: str,
        tx: SafeTransaction,
        tx_hash: str,
        owners: List[str],
        threshold: int,
    ) -> TransactionRequest:
        """
        Create a new signature collection request.

        Args:
            safe_address: Address of the Safe.
            tx: The Safe transaction.
            tx_hash: Transaction hash for signing.
            owners: List of owner addresses.
            threshold: Number of signatures required.

        Returns:
            New TransactionRequest.
        """
        with self._lock:
            now = current_timestamp()
            request = TransactionRequest(
                id=generate_request_id(),
                safe_address=safe_address,
                transaction=tx,
                transaction_hash=tx_hash,
                signatures={},
                threshold=threshold,
                created_at=now,
                expires_at=now + self._expiration_seconds,
            )
            self._pending_requests[request.id] = request
            return request

    def add_signature(self, request_id: str, sig: SafeSignature) -> None:
        """
        Add a signature to a request.

        Args:
            request_id: Request ID.
            sig: Signature to add.

        Raises:
            ValueError: If request not found, expired, or already signed.
        """
        with self._lock:
            request = self._pending_requests.get(request_id)
            if request is None:
                raise ValueError("Request not found")

            # Check expiration
            if current_timestamp() > request.expires_at:
                del self._pending_requests[request_id]
                raise ValueError("Request expired")

            # Check if already signed by this signer
            if sig.signer.lower() in request.signatures:
                raise ValueError("Already signed by this signer")

            request.signatures[sig.signer.lower()] = sig

    def get_request(self, request_id: str) -> Optional[TransactionRequest]:
        """
        Get a pending request.

        Args:
            request_id: Request ID.

        Returns:
            TransactionRequest or None if not found/expired.
        """
        with self._lock:
            request = self._pending_requests.get(request_id)
            if request is None:
                return None

            # Check expiration
            if current_timestamp() > request.expires_at:
                del self._pending_requests[request_id]
                return None

            return request

    def remove_request(self, request_id: str) -> bool:
        """
        Remove a request.

        Args:
            request_id: Request ID.

        Returns:
            True if removed, False if not found.
        """
        with self._lock:
            if request_id in self._pending_requests:
                del self._pending_requests[request_id]
                return True
            return False

    def get_pending_requests(self) -> List[TransactionRequest]:
        """
        Get all non-expired pending requests.

        Returns:
            List of pending requests.
        """
        with self._lock:
            self._cleanup_expired()
            return list(self._pending_requests.values())

    def get_pending_owners(
        self, request_id: str, owners: List[str]
    ) -> Optional[List[str]]:
        """
        Get owners who haven't signed a request yet.

        Args:
            request_id: Request ID.
            owners: List of all owner addresses.

        Returns:
            List of pending owner addresses, or None if request not found.
        """
        with self._lock:
            request = self._pending_requests.get(request_id)
            if request is None:
                return None

            pending = []
            for owner in owners:
                if owner.lower() not in request.signatures:
                    pending.append(owner)
            return pending

    def get_signed_owners(self, request_id: str) -> Optional[List[str]]:
        """
        Get owners who have signed a request.

        Args:
            request_id: Request ID.

        Returns:
            List of signed owner addresses, or None if request not found.
        """
        with self._lock:
            request = self._pending_requests.get(request_id)
            if request is None:
                return None

            return list(request.signatures.keys())

    def get_combined_signature(self, request_id: str) -> bytes:
        """
        Get the packed signatures for execution.

        Args:
            request_id: Request ID.

        Returns:
            Combined signature bytes.

        Raises:
            ValueError: If request not found or not ready.
        """
        with self._lock:
            request = self._pending_requests.get(request_id)
            if request is None:
                raise ValueError("Request not found")

            if not request.is_ready():
                raise ValueError("Not enough signatures")

            return combine_signatures(request.signatures)

    def cleanup(self) -> None:
        """Remove expired requests."""
        with self._lock:
            self._cleanup_expired()

    def _cleanup_expired(self) -> None:
        """Remove expired requests (must hold lock)."""
        now = current_timestamp()
        expired = [
            rid
            for rid, req in self._pending_requests.items()
            if now > req.expires_at
        ]
        for rid in expired:
            del self._pending_requests[rid]

    def clear(self) -> None:
        """Remove all pending requests."""
        with self._lock:
            self._pending_requests.clear()
